scio-core/src/main/java/com/spotify/scio/transforms/BaseAsyncLookupDoFn.java

```diff
@@ -61,6 +61,7 @@
   // Data structures for handling async requests
   private final Semaphore semaphore;
   private final ConcurrentMap<UUID, F> futures = new ConcurrentHashMap<>();
+  private final ConcurrentMap<A, F> inFlightRequests = new ConcurrentHashMap<>();
```

Comment:
if I remember correctly a runner might actually create multiple cloned copies of a DoFn and in this case inFlight requests dedup would be scoped only to a single instance of a DoFn bc these DoFns wouldn't share their in flight requests. I wonder if this is somehow should be changed to static like `cache` defined above

Reply:
That'd work, but in the face of a failure we'd be propagating it to all threads and potentially failing multiple bundles.

Reply:
ah, that's a good point. The choice might depend on a specific use case. Would be nice to have this as an option (e.g. `dedupPerWorkerThread`) exposed to users. Not saying it should be added in this PR, just thinking about possible solutions.

Reply:
This is a good question. In fact, looking at the cache, it's seems to be duplicated per-instance... I wonder if we would get comparable performance benefits from just making the _cache_ [per-class](https://github.com/spotify/scio/blob/d54ffd473501bcb610502636784676a9ecf4dd28/scio-core/src/main/java/com/spotify/scio/transforms/DoFnWithResource.java#L73) rather than per-instance (or in line with @stormy-ua's suggestion, making it configurable by the user)

Reply:
I don't think per-class caching works since there can be various subclasses and static fields in abstract classes are shared across all subclasses.

Now thinking that we may want to make this opt-in since users may expect each processed item to side-effect in the external system.

Reply:
yeah I was thinking about that; we could possibly work around it. for example, instead of using  `com.spotify.scio.transforms.BaseAsyncLookupDoFn` as the per-class cache key, using something like [CallSites#getCurrentName](https://github.com/spotify/scio/blob/be7166c58450bd59ae5d0048d39ef3ea0d5ed107/scio-core/src/main/scala/com/spotify/scio/util/CallSites.scala#L76) to use file+line number as a key. that way the cache would be unique across different async lookup transforms, but still shared across all instances+clones of that specific transform.

that's a good point about making this opt-in 👍

Reply:
Does that provide any different functionality than what we're offering now? I feel like I'm probably missing something...